heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java

```
@@ -0,0 +1,419 @@
+                resource);
+        // Insert it into the map
+        instancePlanMap.put(instanceId, instancePlan);
+      }
+
+      containerCpu += (paddingPercentage * containerCpu) / 100;
+      containerRam += (paddingPercentage * containerRam) / 100;
+      containerDiskInBytes += (paddingPercentage * containerDiskInBytes) / 100;
+
+      PackingPlan.Resource resource =
+          new PackingPlan.Resource(Math.round(containerCpu), containerRam, containerDiskInBytes);
+
+      PackingPlan.ContainerPlan containerPlan =
+          new PackingPlan.ContainerPlan(containerId, instancePlanMap, resource);
+
+      containerPlanMap.put(containerId, containerPlan);
+      topologyRam += containerRam;
+      topologyCpu += Math.round(containerCpu);
+      topologyDisk += containerDiskInBytes;
+    }
+
+    // Take the heron internal container into account and the application master for YARN
+    // scheduler
+    topologyRam += instanceRamDefault;
+    topologyDisk += instanceDiskDefault;
+    topologyCpu += instanceCpuDefault;
+
+    PackingPlan.Resource resource = new PackingPlan.Resource(
+        topologyCpu, topologyRam, topologyDisk);
+
+    PackingPlan plan = new PackingPlan(topology.getId(), containerPlanMap, resource);
+    return plan;
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  /**
+   * Get the RAM requirements of all the components
+   *
+   * @return The list of components and their RAM requirements
+   */
+  protected ArrayList<RamRequirement> getRAMInstances() {
+    ArrayList<RamRequirement> ramRequirements = new ArrayList<>();
+    Map<String, Integer> parallelismMap = TopologyUtils.getComponentParallelism(topology);
+    Map<String, Long> ramMap = TopologyUtils.getComponentRamMapConfig(topology);
+
+    for (String component : parallelismMap.keySet()) {
+      if (ramMap.containsKey(component)) {
+        if (!isValidInstance(ramMap.get(component), instanceCpuDefault, instanceDiskDefault)) {
+          return null;
```

what does returning null mean? If it's an error condition can we throw an exception?